### PR TITLE
feat: nostr, onboarding, user-data base

### PR DESCRIPTION
**Pull request by ice-orion**
### What does this PR do?
This PR account initialization, onboarding base flow

### Changes brought by this PR
* Fix onboarding steps UI
* Add router redirects to onboarding when needed
* Add onboarding providers
* Add nostr keystore provider + secured persistance
* Add UserDelegation, UserMetadata, UserIdentity models and providers

### Additional information
https://github.com/user-attachments/assets/c9c9f5c9-7579-42fb-a66e-d555a9fd1046

### CODE REVIEW

**Review by ice-alcides — APPROVED** [review]

File: lib/app/router/providers/go_router_provider.dart [hunk 1]
```diff
@@ -60,3 +47,30 @@ GoRouter goRouter(GoRouterRef ref) {
     navigatorKey: rootNavigatorKey,
   );
 }
+
+FutureOr<String?> _mainRedirect({
```
Comment (ice-tychon) [inline lib/app/router/providers/go_router_provider.dart:51]: Consider make it public and provide tests for it.

File: lib/app/features/wallets/providers/main_wallet_provider.dart [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+part 'main_wallet_provider.g.dart';
+
+@Riverpod(keepAlive: true)
+Future<Wallet?> mainWallet(MainWalletRef ref) async {
+  final currentIdentityKeyName = ref.watch(currentIdentityKeyNameSelectorProvider);
+  if (currentIdentityKeyName == null) {
+    return null;
+  }
+  // TODO: take the list from walletsDataNotifierProvider when connected to ionClient
+  final ionClient = await ref.read(ionApiClientProvider.future);
```
Comment (ice-tychon) [inline lib/app/features/wallets/providers/main_wallet_provider.dart:19]: Should be `ref.watch`

File: lib/app/features/wallets/providers/main_wallet_provider.dart [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    network: mainWallet.network,
+    address: mainWallet.address,
+    name: mainWallet.name,
+    status: mainWallet.status,
+    signingKey: WalletSigningKey(
+      scheme: mainWallet.signingKey.scheme,
+      curve: mainWallet.signingKey.curve,
+      publicKey: mockedTonWalletKeystore.publicKey,
+    ),
+  );
```
Comment (ice-tychon) [inline lib/app/features/wallets/providers/main_wallet_provider.dart:38]:
```suggestion
    final signingKey = mainWallet.signingKey;
  return mainWallet.copyWith(
    signingKey: signingKey.copyWith(publicKey: mockedTonWalletKeystore.publicKey),
  );
```

File: lib/app/features/user/providers/user_relays_provider.dart [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import 'package:ice/app/features/user/providers/current_user_indexers_provider.dart';
+import 'package:nostr_dart/nostr_dart.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'user_relays_provider.g.dart';
+
+@Riverpod(keepAlive: true)
+class UsersRelaysStorage extends _$UsersRelaysStorage {
+  @override
+  Map<String, UserRelays> build() {
```
Comment (ice-tychon) [inline lib/app/features/user/providers/user_relays_provider.dart:16]: Why is it not a family-provider?

File: lib/app/features/user/providers/user_relays_provider.dart [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+@Riverpod(keepAlive: true)
+Future<UserRelays?> userRelays(UserRelaysRef ref, String pubkey) async {
+  final userRelays = ref.watch(usersRelaysStorageProvider.select((state) => state[pubkey]));
+  if (userRelays != null) {
+    return userRelays;
+  }
+
+  final currentUserIndexers = await ref.read(currentUserIndexersProvider.future);
```
Comment (ice-tychon) [inline lib/app/features/user/providers/user_relays_provider.dart:50]: should be `ref.watch`

File: lib/app/features/user/providers/user_relays_provider.dart [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+  if (currentUserIndexers == null) {
+    throw Exception('Current user indexers are not found');
+  }
+
+  final relay = await ref.read(relayProvider(currentUserIndexers.random).future);
```
Comment (ice-tychon) [inline lib/app/features/user/providers/user_relays_provider.dart:56]: `ref.watch`

**Review by ice-kreios — APPROVED** [review]

File: lib/app/features/auth/providers/onboarding_data_provider.dart [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    final userIdentity = await ref.read(currentUserIdentityProvider.future);
+    if (userIdentity == null) {
+      throw Exception('User identity is null');
+    }
+
+    final (:name, :displayName, :languages, :followees) = requireValues();
+
+    final relayUrls = await _assignUserRelays(userIdentity: userIdentity, followees: followees);
```
Comment (ice-tychon) [inline lib/app/features/auth/providers/onboarding_data_provider.dart:99]:
```suggestion
        final (relayUrls, nostrKeyStore) = await (
      _assignUserRelays(userIdentity: userIdentity, followees: followees),
      _generateNostrKeyStore(identityKeyName: currentIdentityKeyName)
    ).wait;
    await (
      _updateUserRelays(nostrKeyStore: nostrKeyStore, relayUrls: relayUrls),
      _updateMetadata(nostrKeyStore: nostrKeyStore, name: name, displayName: displayName),
      _updateLanguages(languages: languages),
      _updateFollowees(followees: followees),
      _updateDelegation(nostrKeyStore: nostrKeyStore)
    ).wait;
```

File: lib/app/features/auth/providers/onboarding_data_provider.dart [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+  const factory OnboardingState({
+    String? name,
+    String? displayName,
+    List<String>? languages,
+    List<String>? followees,
+  }) = _OnboardingState;
+}
+
+@Riverpod(keepAlive: true)
+class OnboardingData extends _$OnboardingData {
```
Comment (ice-tychon) [inline lib/app/features/auth/providers/onboarding_data_provider.dart:32]: This class is waaay too large, and gonna be even larger. Consider splitting it into smaller ones.

**Review by ice-tychon — DISMISSED** [review]

**Review by ice-kreios — APPROVED** [review]

**Review by ice-tychon — APPROVED** [review]

**Review by ice-alcides — APPROVED** [review]

**Pull request merged**